cpp/src/arrow/timestamp_converter.h

```diff
@@ -0,0 +1,41 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+#pragma once
+
+#include <memory>
+
+#include "arrow/type.h"
+
+namespace arrow {
+class TimestampConverter {
+ public:
+  using value_type = TimestampType::c_type;
+  virtual bool operator()(const std::shared_ptr<DataType>& type, const char* s,
+                          size_t length, value_type* out) const = 0;
+};
+
+class StrptimeConverter : public TimestampConverter {
+ private:
+  std::string format_;
```

Comment:
This converter need not own the format string; use `util::string_view format_;` instead

Reply:
We need to decide how to deal with interfaces before.

Reply:
The format string should be small, so this is not a large issue.

This class should probably be called `StrptimeTimestampParser` or something to more clearly indicate. It will need a docstring so that users know where to find the grammar supported by the format strings